#### **Problem Statement:**

1.  The main objective of the problem statement is to predict the next word from the imdb movie review dataset by applying LSTM.

2.  We need to perform the LSTM and the sentimental analysis to get the next word.


#### **Step 1: Import all the necessary libraries**

In [1]:
import numpy              as np
import pandas             as pd
import matplotlib.pyplot  as plt
import seaborn            as sns

from   tensorflow.keras.datasets import imdb
from   tensorflow.keras.models   import Sequential
from   tensorflow.keras.layers   import Dense, LSTM, Embedding, Flatten
from   tensorflow.keras.utils    import to_categorical

#### **OBSERVATIONS:**

1.  imdb ------> It is an imdb movie review dataset that has 50000 labels

2.  Sequential ------> It has a sequential stack of layers after layers

3.  Dense ---------> It forms a dense connected layer

4.  LSTM ------->  It learns the sequential relationship and remembers the information for a longer period of time.

5.  Embedding -------> It converts the words into the dense dimensional vectors.

6.  Flatten ---------> converts the 2d or 3d feature maps into 1d vectors.

7.  to_categorical -------> It converts the integer labels into the dense vectors.

#### **Step 2: Load the imdb movie review dataset**

In [2]:
#### Load the imdb movie review dataset

#### define the vocabulary size
vocab_size=10000
(X_train,Y_train),(X_test, Y_test)=imdb.load_data(num_words=vocab_size)

In [3]:
X_train

array([list([1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4, 173, 36, 256, 5, 25, 100, 43, 838, 112, 50, 670, 2, 9, 35, 480, 284, 5, 150, 4, 172, 112, 167, 2, 336, 385, 39, 4, 172, 4536, 1111, 17, 546, 38, 13, 447, 4, 192, 50, 16, 6, 147, 2025, 19, 14, 22, 4, 1920, 4613, 469, 4, 22, 71, 87, 12, 16, 43, 530, 38, 76, 15, 13, 1247, 4, 22, 17, 515, 17, 12, 16, 626, 18, 2, 5, 62, 386, 12, 8, 316, 8, 106, 5, 4, 2223, 5244, 16, 480, 66, 3785, 33, 4, 130, 12, 16, 38, 619, 5, 25, 124, 51, 36, 135, 48, 25, 1415, 33, 6, 22, 12, 215, 28, 77, 52, 5, 14, 407, 16, 82, 2, 8, 4, 107, 117, 5952, 15, 256, 4, 2, 7, 3766, 5, 723, 36, 71, 43, 530, 476, 26, 400, 317, 46, 7, 4, 2, 1029, 13, 104, 88, 4, 381, 15, 297, 98, 32, 2071, 56, 26, 141, 6, 194, 7486, 18, 4, 226, 22, 21, 134, 476, 26, 480, 5, 144, 30, 5535, 18, 51, 36, 28, 224, 92, 25, 104, 4, 226, 65, 16, 38, 1334, 88, 12, 16, 283, 5, 16, 4472, 113, 103, 32, 15, 16, 5345, 19, 178, 32]),
       list([1, 194, 1153, 194, 8255, 78, 228,

In [4]:
Y_train

array([1, 0, 0, ..., 0, 1, 0])

In [5]:
X_test

array([list([1, 591, 202, 14, 31, 6, 717, 10, 10, 2, 2, 5, 4, 360, 7, 4, 177, 5760, 394, 354, 4, 123, 9, 1035, 1035, 1035, 10, 10, 13, 92, 124, 89, 488, 7944, 100, 28, 1668, 14, 31, 23, 27, 7479, 29, 220, 468, 8, 124, 14, 286, 170, 8, 157, 46, 5, 27, 239, 16, 179, 2, 38, 32, 25, 7944, 451, 202, 14, 6, 717]),
       list([1, 14, 22, 3443, 6, 176, 7, 5063, 88, 12, 2679, 23, 1310, 5, 109, 943, 4, 114, 9, 55, 606, 5, 111, 7, 4, 139, 193, 273, 23, 4, 172, 270, 11, 7216, 2, 4, 8463, 2801, 109, 1603, 21, 4, 22, 3861, 8, 6, 1193, 1330, 10, 10, 4, 105, 987, 35, 841, 2, 19, 861, 1074, 5, 1987, 2, 45, 55, 221, 15, 670, 5304, 526, 14, 1069, 4, 405, 5, 2438, 7, 27, 85, 108, 131, 4, 5045, 5304, 3884, 405, 9, 3523, 133, 5, 50, 13, 104, 51, 66, 166, 14, 22, 157, 9, 4, 530, 239, 34, 8463, 2801, 45, 407, 31, 7, 41, 3778, 105, 21, 59, 299, 12, 38, 950, 5, 4521, 15, 45, 629, 488, 2733, 127, 6, 52, 292, 17, 4, 6936, 185, 132, 1988, 5304, 1799, 488, 2693, 47, 6, 392, 173, 4, 2, 4378, 270, 2352, 4, 1500, 7, 

In [6]:
Y_test

array([0, 1, 1, ..., 0, 0, 0])

#### **OBSERVATIONS:**

1.  The vocabulary size of 10000 words is defined that has all the unique words in it.

2.  Then the imdb movie review dataset is loaded with 10000 most frequent vocabularies.

3.  The imdb movie review dataset is divided into the training and testing data.

#### **Input Data**

In [7]:
print("Shape of the input training data is:", X_train.shape)
print("Shape of the input testing  data is:", X_test.shape)

Shape of the input training data is: (25000,)
Shape of the input testing  data is: (25000,)


#### **Output Data**

In [8]:
print("Shape of the output training data is:", Y_train.shape)
print("Shape of the output testing  data is:", Y_test.shape)

Shape of the output training data is: (25000,)
Shape of the output testing  data is: (25000,)


#### **OBSERVATIONS:**

1.  The imdb movie review dataset has 25000 training labels and 25000 testing labels.

#### **Step 3: Generate the Input Training Sequences**

In [9]:
X_train[:1000]

array([list([1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4, 173, 36, 256, 5, 25, 100, 43, 838, 112, 50, 670, 2, 9, 35, 480, 284, 5, 150, 4, 172, 112, 167, 2, 336, 385, 39, 4, 172, 4536, 1111, 17, 546, 38, 13, 447, 4, 192, 50, 16, 6, 147, 2025, 19, 14, 22, 4, 1920, 4613, 469, 4, 22, 71, 87, 12, 16, 43, 530, 38, 76, 15, 13, 1247, 4, 22, 17, 515, 17, 12, 16, 626, 18, 2, 5, 62, 386, 12, 8, 316, 8, 106, 5, 4, 2223, 5244, 16, 480, 66, 3785, 33, 4, 130, 12, 16, 38, 619, 5, 25, 124, 51, 36, 135, 48, 25, 1415, 33, 6, 22, 12, 215, 28, 77, 52, 5, 14, 407, 16, 82, 2, 8, 4, 107, 117, 5952, 15, 256, 4, 2, 7, 3766, 5, 723, 36, 71, 43, 530, 476, 26, 400, 317, 46, 7, 4, 2, 1029, 13, 104, 88, 4, 381, 15, 297, 98, 32, 2071, 56, 26, 141, 6, 194, 7486, 18, 4, 226, 22, 21, 134, 476, 26, 480, 5, 144, 30, 5535, 18, 51, 36, 28, 224, 92, 25, 104, 4, 226, 65, 16, 38, 1334, 88, 12, 16, 283, 5, 16, 4472, 113, 103, 32, 15, 16, 5345, 19, 178, 32]),
       list([1, 194, 1153, 194, 8255, 78, 228,

In [10]:
#### define the list for empty sequences
input_sequences = []
for review in X_train[:1000]:
  for i in range(1,len(review)):
    n_gram = review[:i+1]
    #### append all the sequences into the input sequences
    input_sequences.append(n_gram)

In [11]:
input_sequences

[[1, 14],
 [1, 14, 22],
 [1, 14, 22, 16],
 [1, 14, 22, 16, 43],
 [1, 14, 22, 16, 43, 530],
 [1, 14, 22, 16, 43, 530, 973],
 [1, 14, 22, 16, 43, 530, 973, 1622],
 [1, 14, 22, 16, 43, 530, 973, 1622, 1385],
 [1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65],
 [1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458],
 [1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468],
 [1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66],
 [1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941],
 [1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4],
 [1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4, 173],
 [1,
  14,
  22,
  16,
  43,
  530,
  973,
  1622,
  1385,
  65,
  458,
  4468,
  66,
  3941,
  4,
  173,
  36],
 [1,
  14,
  22,
  16,
  43,
  530,
  973,
  1622,
  1385,
  65,
  458,
  4468,
  66,
  3941,
  4,
  173,
  36,
  256],
 [1,
  14,
  22,
  16,
  43,
  530,
  973,
  1622,
  1385,
  65,
  458,
  4468,
  66,
  3941,
  4,
  173,
  36,
  256,
  

#### **OBSERVATIONS:**

1.  Here all the input training sequences are generated for the training data.

#### **Step 4: Perform Padding on all the input sequences**

In [12]:
#### define the maximum length of the input sequence
max_length = max([len(sequence) for sequence in input_sequences])

In [13]:
max_length

1038

In [14]:
#### Applying the padding on the input sequences(apply pre-padding)
from tensorflow.keras.preprocessing.sequence import pad_sequences

input_sequences = pad_sequences(
    input_sequences                                                               ,
    maxlen                                =              max_length               ,
    padding                               =              'pre'
)

In [15]:
input_sequences

array([[   0,    0,    0, ...,    0,    1,   14],
       [   0,    0,    0, ...,    1,   14,   22],
       [   0,    0,    0, ...,   14,   22,   16],
       ...,
       [   0,    0,    0, ...,  624,   25,   28],
       [   0,    0,    0, ...,   25,   28, 1678],
       [   0,    0,    0, ...,   28, 1678,   72]], dtype=int32)

#### **OBSERVATIONS:**

1.  Initially all the input sequences are of the variable length.

2.  Padding has been applied on the input sequences to make all the input sequences with the same length of 1851.

3.  Each and every word id has been converted into the dense dimensional vectors with the length of 1851.

4.  Here pre-padding has been applied on the input sequence where zero has been applied at the beginning of the input sequences in case of any variable length.

#### **Step 5: Split the input sequences into inputs and outputs**

In [16]:
#### define the input sequence

input_sequences

array([[   0,    0,    0, ...,    0,    1,   14],
       [   0,    0,    0, ...,    1,   14,   22],
       [   0,    0,    0, ...,   14,   22,   16],
       ...,
       [   0,    0,    0, ...,  624,   25,   28],
       [   0,    0,    0, ...,   25,   28, 1678],
       [   0,    0,    0, ...,   28, 1678,   72]], dtype=int32)

In [17]:
#### define the input
X=input_sequences[:,:-1]

In [18]:
X

array([[   0,    0,    0, ...,    0,    0,    1],
       [   0,    0,    0, ...,    0,    1,   14],
       [   0,    0,    0, ...,    1,   14,   22],
       ...,
       [   0,    0,    0, ..., 1678,  624,   25],
       [   0,    0,    0, ...,  624,   25,   28],
       [   0,    0,    0, ...,   25,   28, 1678]], dtype=int32)

In [19]:
#### define the output
Y=input_sequences[:,-1]

In [20]:
Y

array([  14,   22,   16, ...,   28, 1678,   72], dtype=int32)

#### **OBSERVATIONS:**

1.  The input sequences has been divided into the input and the output.

In [21]:
#### Convert the output to categorical
Y = to_categorical(Y,num_classes=vocab_size)

In [22]:
Y

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [23]:
#### Check the shapes
print("Shape of the input",  X.shape)
print("Shape of the output", Y.shape)

Shape of the input (242901, 1037)
Shape of the output (242901, 10000)


#### **OBSERVATIONS:**

1.  The input data has 242901 rows and 1037 columns becuase the maximum length of the input sequence is 1037.

2.  The output data has 24201 rows and 10000 columns as the vocabulary size is 10000.

#### **Step 6:  Architecture of LSTM**



  Input Sequence

        ↓

    Embedding

        ↓
       LSTM

        ↓

      Dense

        ↓

    Next Word

In [24]:
#### Construct the model
model = Sequential([
    #### create an embedding layer
    Embedding(input_dim=vocab_size, output_dim=100, input_length = max_length)                         ,
    #### create an LSTM Layer
    LSTM(150)                                                                                          ,
    #### Create an output layer
    Dense(units=vocab_size, activation='softmax')
])

/usr/local/lib/python3.12/dist-packages/keras/src/layers/core/embedding.py:100: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


#### **OBSERVATIONS:**

1.  The sequential model is created with the following layers:--

    -------------------------Emdedding---------------------------------------

    (a.)  The embedding layer is created where we are considering 10000 most frequent words.

    (b.)  Each word has a word id. Then each word id is converted into 100 dense dimensional vectors where each vector is of the length 1038

    (c.) The main purpose of the embedding layer is to convert the word id into the dense dimensional vectors.

     -------------------------LSTM------------------------------------------

     (a.)  LSTM remembers the input sequences for a longer period of time.

     (b.)  It solves the issue of Vanishing gradient descent.

     (c.)  It works better than the SimpleRNN

     (d.)  This layer is applied over 150 neurons where it remembers the sequences of the past events and predicts the future events.

     -----------------------Output-------------------------------------------

     (a.) Here the output layer is created with 10000 neurons and softmax activation fuction is applied over each neuron to get the probability of each word.



In [25]:
model

<Sequential name=sequential, built=False>

#### **Step 7: Get the summary of the model**

In [26]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding (Embedding)           │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ lstm (LSTM)                     │ ?                      │   0 (unbuilt) │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ ?                      │   0 (unbuilt) │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 0 (0.00 B)

 Trainable params: 0 (0.00 B)

 Non-trainable params: 0 (0.00 B)

#### **OBSERVATIONS:**

1.  The above table provides the summary of the model with the given layers

      (a.)  embedding layer

      (b.)  lstm layer

      (c.)  dense layer

#### **Step 8: Compile the model**

In [27]:
#### Compile the model
model.compile(
    optimizer   =   'adam'                          ,
    loss        =   'categorical_crossentropy'      ,
    metrics     =   ['accuracy']
)

#### **OBSERVATIONS:**

1.  The model has been made ready for training with the following parameters :-

      (a.)  optimizer = 'adam'. It updates the weights during the time of backpropogation

      (b.)  loss = 'categorical_crossentropy'. It solves the multi classification problem.

      (c.)  metrics = 'accuracy'. It measures the performance of the model.

#### **Step 9: Train the model**

In [28]:
#### perform the training on a small data
X = X[:1000]
Y = Y[:1000]

In [29]:
#### Train the model
history = model.fit(
    X                                                    ,
    Y                                                    ,
    epochs              =             5                  ,
    batch_size          =             16                 ,
    validation_split    =             0.2                ,
    verbose             =              1
)

Epoch 1/5
50/50 ━━━━━━━━━━━━━━━━━━━━ 75s 1s/step - accuracy: 0.0562 - loss: 8.1112 - val_accuracy: 0.0700 - val_loss: 6.7020
Epoch 2/5
50/50 ━━━━━━━━━━━━━━━━━━━━ 71s 1s/step - accuracy: 0.0675 - loss: 5.9331 - val_accuracy: 0.0700 - val_loss: 6.6316
Epoch 3/5
50/50 ━━━━━━━━━━━━━━━━━━━━ 72s 1s/step - accuracy: 0.0550 - loss: 5.5561 - val_accuracy: 0.0700 - val_loss: 6.7512
Epoch 4/5
50/50 ━━━━━━━━━━━━━━━━━━━━ 85s 2s/step - accuracy: 0.0675 - loss: 5.4427 - val_accuracy: 0.0700 - val_loss: 6.8937
Epoch 5/5
50/50 ━━━━━━━━━━━━━━━━━━━━ 75s 2s/step - accuracy: 0.0675 - loss: 5.3863 - val_accuracy: 0.0700 - val_loss: 6.8168


In [30]:
history

#### **OBSERVATIONS:**

1.  The model has been trained with the following parameters:--

    (a.)  X --------->  Input sequence

    (b.)  Y --------->  Output data

    (c.)  epochs = 10 ---------->  No of iterations needed to train the model

    (d.)  batch_size = 64 ------> Instead of processing all the records in one iteration, in each iteration, it is considering 128 records per batch and processing the records in batch wise manner

    (e.)  validation_split = 0.2 -------> 20% of the data is meant for validation and helps the model to know how well it performs on the unseen data

    (f.)  verbose = 1 ----------->  It controls the training output.

2.  The accuracy of the model increases over the epochs and the loss decreases over the epochs.

#### **Step 10: Evaluate the model**

In [31]:
loss, accuracy = model.evaluate(X,Y,verbose=0)
print(f"Loss: {loss:.3f}")
print(f"Accuracy: {accuracy:.3f}")

Loss: 5.553
Accuracy: 0.068


#### **OBSERVATIONS:**

1.  After evaluating the model, the loss is very high but the accuracy is less.

#### **Step 11: Create Reverse Dictionary**

In [32]:
#### get the word index
word_index = imdb.get_word_index()
print(word_index)

#### get the reverse word index
reverse_word_index = {value+3:key for key,value in word_index.items()}
print(reverse_word_index)

reverse_word_index[0] = "<PAD>"
reverse_word_index[1] = "<START>"
reverse_word_index[2] = "<UNK>"

{'fawn': 34701, 'tsukino': 52006, 'nunnery': 52007, 'sonja': 16816, 'vani': 63951, 'woods': 1408, 'spiders': 16115, 'hanging': 2345, 'woody': 2289, 'trawling': 52008, "hold's": 52009, 'comically': 11307, 'localized': 40830, 'disobeying': 30568, "'royale": 52010, "harpo's": 40831, 'canet': 52011, 'aileen': 19313, 'acurately': 52012, "diplomat's": 52013, 'rickman': 25242, 'arranged': 6746, 'rumbustious': 52014, 'familiarness': 52015, "spider'": 52016, 'hahahah': 68804, "wood'": 52017, 'transvestism': 40833, "hangin'": 34702, 'bringing': 2338, 'seamier': 40834, 'wooded': 34703, 'bravora': 52018, 'grueling': 16817, 'wooden': 1636, 'wednesday': 16818, "'prix": 52019, 'altagracia': 34704, 'circuitry': 52020, 'crotch': 11585, 'busybody': 57766, "tart'n'tangy": 52021, 'burgade': 14129, 'thrace': 52023, "tom's": 11038, 'snuggles': 52025, 'francesco': 29114, 'complainers': 52027, 'templarios': 52125, '272': 40835, '273': 52028, 'zaniacs': 52130, '275': 34706, 'consenting': 27631, 'snuggled': 408

#### **Step 12: Predict Next Word**

In [33]:
def predict_next_word(seed_text):

    token_list = [
        word_index.get(word,2)+3
        for word in seed_text.lower().split()
    ]

    token_list = pad_sequences(
        [token_list],
        maxlen=max_length-1,
        padding='pre'
    )

    predicted = model.predict(
        token_list,
        verbose=0
    )

    predicted_word_id = np.argmax(predicted)

    return reverse_word_index.get(
        predicted_word_id,
        "Unknown"
    )

#### **Step 13: Test Model**

In [34]:
seed_text = "this movie"

print(
    "Next Word :",
    predict_next_word(seed_text)
)

Next Word : the
